features/credentials/provider_credentials_controller.rb

```diff
@@ -0,0 +1,91 @@
+require "set"
+
+require_relative "../../shared/authenticated_controller_base"
+require_relative "../../services/services"
+
+module FastlaneCI
+  # A CRUD controller to manage provider credentials associated with a user. A
+  # `ProviderCredential` is a credential a user may use to access some third-party
+  # provider. For instance, the `GitHubProviderCredential` allows FastlaneCI users
+  # to interact with the GitHub API
+  class ProviderCredentialsController < AuthenticatedControllerBase
+    HOME = "/provider_credentials"
+
+    get HOME do
+      locals = { title: "Provider Credentials" }
+      erb(:provider_credentials, locals: locals, layout: FastlaneCI.default_layout)
+    end
+
+    post "#{HOME}/create" do
+      if valid_params?(params, provider_credential_params) &&
+         valid_user_id?(params[:user_id])
+        Services.provider_credential_service.create_provider_credential!(
+          format_params(params, provider_credential_params)
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    post "#{HOME}/update" do
+      if valid_params?(params, provider_credential_params) &&
+         valid_user_id?(params[:user_id])
+        Services.provider_credential_service.update_provider_credential!(
+          format_params(params, provider_credential_params)
+        )
+      end
+
+      redirect(HOME)
+    end
+
+    private
+
+    #####################################################
+    # @!group Data: View-specific data
+    #####################################################
+
+    # @return [Array[User]]
+    def users
+      Services.user_service.user_data_source.users
```

Comment:
services are meant to abstract away the details of their datasources. So for this one, let's not reach in to the service's properties(`user_data_source`), instead, let's expose a method like `def users`